cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaGenerator.ts

```diff
@@ -235,6 +237,12 @@ function SchemaGenerator(schemaTree: INode) {
       if (isFieldComplexType) {
         field.type = generateSchemaFromComplexType(type, currentField, nullable);
       }
+      if (!isEmpty(typeProperties) && typeProperties.doc) {
+        field.doc = typeProperties.doc;
+      }
+      if (!isEmpty(typeProperties) && typeProperties.aliases) {
+        field.aliases = typeProperties.aliases;
+      }
```

Comment:
What are the types of `doc` and `aliases`? `aliases` is an array, correct? Should they be copied here to avoid any issues from downstream mutations?

Reply:
Some of the operations that happens in schema manager does in place changes instead of creating copies. This is purely done for performance. In a schema with 5k-10k fields with 10 levels nesting there was a lot more performance issue with creating copies to avoid mutation. Most of the operations happening in the tree are mutations. 

Specifically here the `doc` is a string and `aliases` is an array of string. We intentionally break the chain between mutations happening within the component to state maintained in schema manager. If you strongly feel we should use ImmutableJS its a broader conversation we should probably discuss outside this PR.

Reply:
I'm not arguing for Immutable.js. I was just wondering if doing a `.slice()` on these arrays when they get assigned would avoid potential issues down the line.